docs/source/blueprints/BP002.rst

```diff
@@ -0,0 +1,220 @@
+            {"type": "vmimage", "distro": "fedora", "version": 31, "arch": "x86_64"},
+            {"type": "package", "package": "lvm"}
+        ]
+
+The requirement `type` should match the module responsible for that type of
+requirement.
+
+Requirements representation as Python executable
+------------------------------------------------
+
+Another way to create the requirements representation as JSON files is by
+writing a Python executable. This approach makes the requirements
+representation flexible, by allowing the use of Python variables and code that
+may change the parameters values for the requirements, depending on the
+environment the Python code runs.
+
+The following example shows a requirement that depends on the architecture
+the test is running::
+
+        #!/usr/bin/python3
+
+        import os
+        import json
+
+        requirements = [
+            {"type": "url", "url": "https://cabort.com/cabort.c", "hash": "deadbeefdeadbeef"},
+            {"type": "vmimage", "distro": "fedora", "version": 31, "arch": os.uname()[4]},
+            {"type": "package", "package": "lvm"}
+        ]
+
+        print(json.dumps(requirements))
+
+Requirements representation as metadata on test docstring
+---------------------------------------------------------
+
+Test writers may want to add the requirements of a test into de test code. The
+option proposed here allows the use of metadata on test docstrings to represent
+the requirements list.
+
+Below is an example of how to define requirements as metadata on docstrings::
+
+        def test_something(self):
+        '''
+            :avocado requirement: [{"type": "url", "url": "https://cabort.com/cabort.c", "hash": "deadbeefdeadbeef"}]
+            :avocado requirement: [{"type": "vmimage", "distro": "fedora", "version": 31, "arch": "x86_64"}]
+            :avocado requirement: [{"type": "package", "package": "lvm"}]
+        '''
+           <test code>
+
+Requirements files location
+===========================
+
+It may be useful for test writers to define a standard source location for the
+requirements JSON files and the requirements Python executable.
+
+This blueprint defines the default location for those files into a requirements
+directory preceded by the test file name. It is also supported the use of
+sub-directories using the name of a specific test to define requirements for
+that test.
+
+The following file tree is an example of possible use for requirements
+directories for a test::
+
+        cabort.py
+        cabort.py.requirements/
+        ├── CAbort.test_2
+        │   └── requirements.py
+        └── requirements.json
+
+In this case, all the tests on `cabort.py`, except for `CAbort.test_2`, use the
+`requirements.json` file located at `cabort.py.requirements`. The `CAbort.test_2`
+test uses its own `requirements.py` located at `CAbort.test_2` directory inside
+the requirements root directory.
```

Comment:
Are we going to have a "job wide" requirements? Let's say that all my tests need the same image. Do I need to specify multiple requirements?

Reply:
I agree we need a directory-wide requirements file, and it is missing here. I remember we discussed that, but I forgot to add. WIll do in the next round.

Reply:
The "job wide" requirements seems similar (use case wise) to what @willianrampazzo proposed with the command line argument.  The use case that I believe would be the most common, though, would be "job wide", but "test suite wide".  Now, bear with me, because we don't have the "test suite" concept defined in Avocado, but what I mean by that is a file structure on a location that also contains multiple tests.  All tests would then also "respect" their "parent" requirements file.

But, even this feature could be initially replaced by solutions outside of Avocado, such as symbolic links.  So, I'd defer this to a second stage, and at the same time, expand the "data" file behavior to match it.